src/window_gui.h

```diff
@@ -248,6 +248,150 @@ DECLARE_ENUM_AS_BIT_SET(WindowFlags)
 static const int TIMEOUT_DURATION = 7; ///< The initial timeout value for WF_TIMEOUT.
 static const int WHITE_BORDER_DURATION = 3; ///< The initial timeout value for WF_WHITE_BORDER.
 
+/**
+ * Cache and iterator for viewport management. The cache is automatically managed by the factory and destructor
+ * of the \c ViewportData struct.
+ *
+ * @see ViewportData::Create()
+ * @see ViewportData::~ViewportData
+ */
+typedef std::vector<std::weak_ptr<ViewPort>> ViewportCacheBase;
```

Comment:
Random thought: Do we actually need the complexity and overhead of weak pointers here, or would it be better to place viewport management under the memory pool system, where we already get iteration (and, I assume, automagic memory management) for free?

Reply:
I don't really see why either of these are necessary. struct ViewPort already has a single owner: struct Window.
Using a std::unique_ptr would make sense.

Resorting to std::shared_ptr creates a lot of unnecessary complexity and overheads. You'd be better off just forgetting about the Viewport pointer cache and iterating the window hierarchy like before.

Reply:
It is true that neither are strictly necessary so long as the program is correct. It's aimed more as a safeguard for that one case where some part of the program takes a (temporary) reference to a viewport and the window owning it gets deleted. Likewise, there should never be a case of a viewport pointer in the cache being invalid; so long as the program can never delete a viewport while another part of it is iteration through them.

Similar story for the viewport overlays: unless it can be proven that nothing can hold a reference to an overlay beyond the lifespan of its owning viewport, it needs to be a shared pointer.

I am happy to use `unique_ptr`s if it makes sense to do so; I was acting upon @michicc's suggestion to move to automatic memory management.